### PR TITLE
net-proxy/yass: add riscv and loong arches

### PR DESCRIPTION
**Pull request by Chilledheart**
image

Cross-compiling tasks are passing including unittest under qemu emulation. It should be safe to add these two architectures: riscv64 and loongarch64.

### CODE REVIEW

**Comment by liangyongxiang** [comment]
I'll merge it after doing a simple running test on real hardware.

**Comment by liangyongxiang** [comment]
build faild in loong arch : https://fars.ee/hWyx
emerge --info : https://fars.ee/G1QB

**Comment by Chilledheart** [comment]
Thanks for the quick reply. It seems emerge picks RelWithDebInfo build type but not Release build type. You might override it with `-DCMAKE_BUILD_TYPE=Release` to continue the test.

In the meantime, I will fix RelWithDebInfo build soon.

**Comment by Chilledheart** [comment]
Updated with patch. It seems it works fine with RelWithDebInfo build type locally (cross-compiling).

**Comment by liangyongxiang** [comment]
I ran it and got an exception.

```
$ yass
$ coredumpctl gdb
....
(gdb) bt
#0  0x00007ffff06918f0 in ??? () at /usr/lib64/libc.so.6
#1  0x00007ffff064b668 in raise () at /usr/lib64/libc.so.6
#2  0x00007ffffeca087c in <signal handler called> ()
#3  0x00007ffff06918f0 in ??? () at /usr/lib64/libc.so.6
#4  0x00007ffff064b668 in raise () at /usr/lib64/libc.so.6
#5  0x00007ffff0636764 in abort () at /usr/lib64/libc.so.6
#6  0x00007fffc8995e98 in llvm::report_fatal_error(llvm::Twine const&, bool) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#7  0x00007fffc9373358 in llvm::SelectionDAGISel::CannotYetSelect(llvm::SDNode*) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#8  0x00007fffc93796cc in llvm::SelectionDAGISel::SelectCodeCommon(llvm::SDNode*, unsigned char const*, unsigned int) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#9  0x00007fffcb6676fc in ??? () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#10 0x00007fffc9371c70 in llvm::SelectionDAGISel::DoInstructionSelection() () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#11 0x00007fffc937c460 in llvm::SelectionDAGISel::CodeGenAndEmitDAG() () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#12 0x00007fffc937def8 in llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#13 0x00007fffc937f9b4 in llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#14 0x00007fffc8e6094c in llvm::MachineFunctionPass::runOnFunction(llvm::Function&) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#15 0x00007fffc8bad240 in llvm::FPPassManager::runOnFunction(llvm::Function&) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#16 0x00007fffc8bad528 in llvm::FPPassManager::runOnModule(llvm::Module&) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#17 0x00007fffc8badb68 in llvm::legacy::PassManagerImpl::run(llvm::Module&) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#18 0x00007fffca91c0c8 in llvm::MCJIT::emitObject(llvm::Module*) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#19 0x00007fffca91c7ec in llvm::MCJIT::generateCodeForModule(llvm::Module*) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#20 0x00007fffca918468 in llvm::MCJIT::finalizeObject() () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#21 0x00007fffca88faa4 in LLVMGetPointerToGlobal () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#22 0x00007fffe612e294 in gallivm_jit_function () at /usr/lib64/dri/swrast_dri.so
#23 0x00007fffe61bebc8 in llvmpipe_update_fs () at /usr/lib64/dri/swrast_dri.so
#24 0x00007fffe61ac738 in llvmpipe_update_derived () at /usr/lib64/dri/swrast_dri.so
#25 0x00007fffe618ad50 in llvmpipe_draw_vbo () at /usr/lib64/dri/swrast_dri.so
#26 0x00007fffe5d9a384 in _mesa_draw_arrays.part.0 () at /usr/lib64/dri/swrast_dri.so
#27 0x00007ffff029d1a4 in gsk_gl_command_queue_execute () at /usr/lib64/libgtk-4.so.1
#28 0x00007ffff02bb744 in gsk_gl_render_job_render () at /usr/lib64/libgtk-4.so.1
#29 0x00007ffff02984b4 in gsk_gl_renderer_render () at /usr/lib64/libgtk-4.so.1
#30 0x00007ffff027b054 in gsk_renderer_render () at /usr/lib64/libgtk-4.so.1
#31 0x00007ffff0088d6c in gtk_widget_render () at /usr/lib64/libgtk-4.so.1
#32 0x00007ffff009064c in surface_render () at /usr/lib64/libgtk-4.so.1
#37 0x00007fffefbb1a40 in <emit signal '???' on instance ???> () at /usr/lib64/libgobject-2.0.so.0
    #33 0x00007ffff0218108 in _gdk_marshal_BOOLEAN__BOXEDv () at /usr/lib64/libgtk-4.so.1
    #34 0x00007fffefb977d8 in _g_closure_invoke_va () at /usr/lib64/libgobject-2.0.so.0
    #35 0x00007fffefbac1d8 in signal_emit_valist_unlocked () at /usr/lib64/libgobject-2.0.so.0
    #36 0x00007fffefbb19bc in g_signal_emit_valist () at /usr/lib64/libgobject-2.0.so.0
#38 0x00007ffff026e8f4 in gdk_surface_paint_on_clock () at /usr/lib64/libgtk-4.so.1
#42 0x00007fffefbb1a40 in <emit signal '???' on instance ???> () at /usr/lib64/libgobject-2.0.so.0
    #39 0x00007fffefb977d8 in _g_closure_invoke_va () at /usr/lib64/libgobject-2.0.so.0
    #40 0x00007fffefbac5cc in signal_emit_valist_unlocked () at /usr/lib64/libgobject-2.0.so.0
    #41 0x00007fffefbb19bc in g_signal_emit_valist () at /usr/lib64/libgobject-2.0.so.0
#43 0x00007ffff025b8fc in gdk_frame_clock_paint_idle () at /usr/lib64/libgtk-4.so.1
#44 0x00007fffefa8dd5c in g_timeout_dispatch () at /usr/lib64/libglib-2.0.so.0
#45 0x00007fffefa8970c in g_main_dispatch () at /usr/lib64/libglib-2.0.so.0
#46 0x00007fffefa8cde8 in g_main_context_iterate_unlocked.isra () at /usr/lib64/libglib-2.0.so.0
#47 0x00007fffefa8d54c in g_main_context_iteration () at /usr/lib64/libglib-2.0.so.0
#48 0x00007fffefcdc7d8 in g_application_run () at /usr/lib64/libgio-2.0.so.0
#49 0x0000555559271958 in YASSApp::ApplicationRun (this=this@entry=0x555575cd9280, argc=argc@entry=1, argv=argv@entry=0x7ffffbb89018)
    at /tmp/portage/net-proxy/yass-1.8.2-r1/work/yass-1.8.2/src/gtk4/yass.cpp:215
#50 0x000055555920dc78 in main (argc=<optimized out>, argv=0x7ffffbb89018) at /tmp/portage/net-proxy/yass-1.8.2-r1/work/yass-1.8.2/src/gtk4/yass.cpp:147
(gdb)
```

**Comment by Chilledheart** [comment]
It seems it crashed in gtk4 and mesa calls. Can you give it a try with other gtk4 applications, or gtk3 applications?

**Comment by Chilledheart** [comment]
I can add another port which runs in CLI mode only. Maybe you can bisect the issue with gtk and yass?

**Comment by liangyongxiang** [comment]
> It seems it crashed in gtk4 and mesa calls. Can you give it a try with other gtk4 applications, or gtk3 applications?

Is there a recommended test program?

**Comment by Chilledheart** [comment]
> > It seems it crashed in gtk4 and mesa calls. Can you give it a try with other gtk4 applications, or gtk3 applications?
> 
> Is there a recommended test program?

Do you mean the gtk4? Maybe you can start with this https://docs.gtk.org/gtk4/getting_started.html.

Create a new file with the following content named example-0.c.
```
#include <gtk/gtk.h>

static void
activate (GtkApplication* app,
          gpointer        user_data)
{
  GtkWidget *window;

  window = gtk_application_window_new (app);
  gtk_window_set_title (GTK_WINDOW (window), "Window");
  gtk_window_set_default_size (GTK_WINDOW (window), 200, 200);
  gtk_window_present (GTK_WINDOW (window));
}

int
main (int    argc,
      char **argv)
{
  GtkApplication *app;
  int status;

  app = gtk_application_new ("org.gtk.example", G_APPLICATION_DEFAULT_FLAGS);
  g_signal_connect (app, "activate", G_CALLBACK (activate), NULL);
  status = g_application_run (G_APPLICATION (app), argc, argv);
  g_object_unref (app);

  return status;
}
```

You can compile the program above with GCC using:

```
gcc $( pkg-config --cflags gtk4 ) -o example-0 example-0.c $( pkg-config --libs gtk4 )
```

**Comment by liangyongxiang** [comment]
> `gcc $( pkg-config --cflags gtk4 ) -o example-0 example-0.c $( pkg-config --libs gtk4 )`

Looks like a problem with my gtk4 environment.

```
[Current thread is 1 (Thread 0x7fffef35c460 (LWP 614232))]
(gdb) bt
#0  0x00007fffefdd58f0 in ??? () at /usr/lib64/libc.so.6
#1  0x00007fffefd8f668 in raise () at /usr/lib64/libc.so.6
#2  0x00007fffefd7a764 in abort () at /usr/lib64/libc.so.6
#3  0x00007fffd0995e98 in llvm::report_fatal_error(llvm::Twine const&, bool) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#4  0x00007fffd1373358 in llvm::SelectionDAGISel::CannotYetSelect(llvm::SDNode*) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#5  0x00007fffd13796cc in llvm::SelectionDAGISel::SelectCodeCommon(llvm::SDNode*, unsigned char const*, unsigned int) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#6  0x00007fffd36676fc in ??? () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#7  0x00007fffd1371c70 in llvm::SelectionDAGISel::DoInstructionSelection() () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#8  0x00007fffd137c460 in llvm::SelectionDAGISel::CodeGenAndEmitDAG() () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#9  0x00007fffd137def8 in llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#10 0x00007fffd137f9b4 in llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#11 0x00007fffd0e6094c in llvm::MachineFunctionPass::runOnFunction(llvm::Function&) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#12 0x00007fffd0bad240 in llvm::FPPassManager::runOnFunction(llvm::Function&) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#13 0x00007fffd0bad528 in llvm::FPPassManager::runOnModule(llvm::Module&) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#14 0x00007fffd0badb68 in llvm::legacy::PassManagerImpl::run(llvm::Module&) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#15 0x00007fffd291c0c8 in llvm::MCJIT::emitObject(llvm::Module*) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#16 0x00007fffd291c7ec in llvm::MCJIT::generateCodeForModule(llvm::Module*) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#17 0x00007fffd2918468 in llvm::MCJIT::finalizeObject() () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#18 0x00007fffd288faa4 in LLVMGetPointerToGlobal () at /usr/lib/llvm/17/lib64/libLLVM-17.so
#19 0x00007fffdf936294 in gallivm_jit_function () at /usr/lib64/dri/swrast_dri.so
#20 0x00007fffdf9c6bc8 in llvmpipe_update_fs () at /usr/lib64/dri/swrast_dri.so
#21 0x00007fffdf9b4738 in llvmpipe_update_derived () at /usr/lib64/dri/swrast_dri.so
#22 0x00007fffdf992d50 in llvmpipe_draw_vbo () at /usr/lib64/dri/swrast_dri.so
#23 0x00007fffdf5a2384 in _mesa_draw_arrays.part.0 () at /usr/lib64/dri/swrast_dri.so
#24 0x00007ffff0ab91a4 in gsk_gl_command_queue_execute () at /usr/lib64/libgtk-4.so.1
#25 0x00007ffff0ad7744 in gsk_gl_render_job_render () at /usr/lib64/libgtk-4.so.1
#26 0x00007ffff0ab44b4 in gsk_gl_renderer_render () at /usr/lib64/libgtk-4.so.1
#27 0x00007ffff0a97054 in gsk_renderer_render () at /usr/lib64/libgtk-4.so.1
#28 0x00007ffff08a4d6c in gtk_widget_render () at /usr/lib64/libgtk-4.so.1
#29 0x00007ffff08ac64c in surface_render () at /usr/lib64/libgtk-4.so.1
#34 0x00007ffff0059a40 in <emit signal '???' on instance ???> () at /usr/lib64/libgobject-2.0.so.0
    #30 0x00007ffff0a34108 in _gdk_marshal_BOOLEAN__BOXEDv () at /usr/lib64/libgtk-4.so.1
    #31 0x00007ffff003f7d8 in _g_closure_invoke_va () at /usr/lib64/libgobject-2.0.so.0
    #32 0x00007ffff00541d8 in signal_emit_valist_unlocked () at /usr/lib64/libgobject-2.0.so.0
    #33 0x00007ffff00599bc in g_signal_emit_valist () at /usr/lib64/libgobject-2.0.so.0
#35 0x00007ffff0a8a8f4 in gdk_surface_paint_on_clock () at /usr/lib64/libgtk-4.so.1
#39 0x00007ffff0059a40 in <emit signal '???' on instance ???> () at /usr/lib64/libgobject-2.0.so.0
    #36 0x00007ffff003f7d8 in _g_closure_invoke_va () at /usr/lib64/libgobject-2.0.so.0
    #37 0x00007ffff00545cc in signal_emit_valist_unlocked () at /usr/lib64/libgobject-2.0.so.0
    #38 0x00007ffff00599bc in g_signal_emit_valist () at /usr/lib64/libgobject-2.0.so.0
#40 0x00007ffff0a778fc in gdk_frame_clock_paint_idle () at /usr/lib64/libgtk-4.so.1
#41 0x00007ffff0a779e0 in gdk_frame_clock_flush_idle () at /usr/lib64/libgtk-4.so.1
#42 0x00007fffeff35d5c in g_timeout_dispatch () at /usr/lib64/libglib-2.0.so.0
#43 0x00007fffeff3170c in g_main_dispatch () at /usr/lib64/libglib-2.0.so.0
#44 0x00007fffeff34de8 in g_main_context_iterate_unlocked.isra () at /usr/lib64/libglib-2.0.so.0
#45 0x00007fffeff3554c in g_main_context_iteration () at /usr/lib64/libglib-2.0.so.0
#46 0x00007ffff01847d8 in g_application_run () at /usr/lib64/libgio-2.0.so.0
#47 0x0000555558ba8ca4 in main ()
(gdb)

```

**Comment by Chilledheart** [comment]
Can you give it a try with gtk3?

Yass is compatible to compile with gtk3 but that needs additional patch.

**Comment by liangyongxiang** [comment]
I ran it and got an exception.

> Yass is compatible to compile with gtk3 but that needs additional patch.

What do I need to do if I want to run it with gtk3?

**Comment by Chilledheart** [comment]
> I ran it and got an exception.
> 
> > Yass is compatible to compile with gtk3 but that needs additional patch.
> 
> What do I need to do if I want to run it with gtk3?

Can you test with gtk3 example at first?

Save the following code as a file named example-0.c:

```
#include <gtk/gtk.h>

static void
activate (GtkApplication* app,
          gpointer        user_data)
{
  GtkWidget *window;

  window = gtk_application_window_new (app);
  gtk_window_set_title (GTK_WINDOW (window), "Window");
  gtk_window_set_default_size (GTK_WINDOW (window), 200, 200);
  gtk_widget_show_all (window);
}

int
main (int    argc,
      char **argv)
{
  GtkApplication *app;
  int status;

  app = gtk_application_new ("org.gtk.example", G_APPLICATION_FLAGS_NONE);
  g_signal_connect (app, "activate", G_CALLBACK (activate), NULL);
  status = g_application_run (G_APPLICATION (app), argc, argv);
  g_object_unref (app);

  return status;
}
```
You can compile the program above with GCC using:
```
gcc `pkg-config --cflags gtk+-3.0` -o example-0 example-0.c `pkg-config --libs gtk+-3.0`
```

**Comment by Chilledheart** [comment]
> I run gtk3 example is ok. 图片

Great. I will upload new patch which makes cmake prefer gtk3 for loongarch64 target.

**Comment by liangyongxiang** [comment]
I'll find some time to check why my 3c5000 can't run the gtk4 program.


I'm now verifying it on riscv as well.

**Comment by liangyongxiang** [comment]
> Uploaded. It will show "Blacklist gtk4 for loongarch64 target" in build log if the patch works as expected.

3c5000：
图片

**Comment by Chilledheart** [comment]
Cool. Let me upload another patch to merge yass-cli target.

**Comment by Chilledheart** [comment]
And one more thing, does gentoo supports risc32 target? I just found out the riscv32 is compatible with yass.
image

**Comment by liangyongxiang** [comment]
> And one more thing, does gentoo supports risc32 target?

I don't think there are any plans for that.

**Comment by Chilledheart** [comment]
> > And one more thing, does gentoo supports risc32 target?
> 
> I don't think there are any plans for that. Because no riscv32 hardware has enough performance to complete the gentoo bootstrap in a reasonable amount of time

Okay. Thanks for explaining.

**Comment by liangyongxiang** [comment]
I also have a lot of packages to upgrade on my 2042, and the real test is going to be rather late. I can merge them first if needed.

**Comment by liangyongxiang** [comment]
> I also have a lot of packages to upgrade on my sg2042, and the real test is going to be rather late. I can merge them first if needed.

**Comment by Chilledheart** [comment]
> I also have a lot of packages to upgrade on my 2042, and the real test is going to be rather late. I can merge them first if needed.

No worry, take your time

**Comment by xen0n** [comment]
> > `gcc $( pkg-config --cflags gtk4 ) -o example-0 example-0.c $( pkg-config --libs gtk4 )`
> 
> Looks like a problem with my gtk4 environment.
> 
> ```
> [Current thread is 1 (Thread 0x7fffef35c460 (LWP 614232))]
> (gdb) bt
> #0  0x00007fffefdd58f0 in ??? () at /usr/lib64/libc.so.6
> #1  0x00007fffefd8f668 in raise () at /usr/lib64/libc.so.6
> #2  0x00007fffefd7a764 in abort () at /usr/lib64/libc.so.6
> #3  0x00007fffd0995e98 in llvm::report_fatal_error(llvm::Twine const&, bool) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
> #4  0x00007fffd1373358 in llvm::SelectionDAGISel::CannotYetSelect(llvm::SDNode*) () at /usr/lib/llvm/17/lib64/libLLVM-17.so
> [snip]
> #19 0x00007fffdf936294 in gallivm_jit_function () at /usr/lib64/dri/swrast_dri.so
> #20 0x00007fffdf9c6bc8 in llvmpipe_update_fs () at /usr/lib64/dri/swrast_dri.so
> #21 0x00007fffdf9b4738 in llvmpipe_update_derived () at /usr/lib64/dri/swrast_dri.so
> #22 0x00007fffdf992d50 in llvmpipe_draw_vbo () at /usr/lib64/dri/swrast_dri.so
> [snip]
> ```

This is https://github.com/loongson-community/discussions/issues/27 -- LoongArch SIMD codegen not yet fully working on LLVM 17. The minimum LLVM version should be 18.

File: net-proxy/yass/yass-1.8.3.ebuild [hunk 1]
```diff
@@ -24,27 +24,27 @@ RDEPEND="
 	sys-libs/zlib
 	net-dns/c-ares
 	net-libs/nghttp2
+	gui? (
+		x11-libs/gtk+:3[wayland?]
+		gui-libs/gtk:4[wayland?]
```
Comment (liangyongxiang) [inline net-proxy/yass/yass-1.8.3.ebuild:29]: Depend on both gtk3 and gtk4?

**Comment by liangyongxiang** [comment]
Also works on sg2042.

File: net-proxy/yass/yass-1.8.3.ebuild (hunk 1 above)
Comment (Chilledheart) [reply net-proxy/yass/yass-1.8.3.ebuild:29]: Yes. It is the best solution I have ever found.

For loongarch64, it depends on gtk3 and for others, it depends on gtk4.

Comment (Chilledheart) [reply net-proxy/yass/yass-1.8.3.ebuild:29]: See https://github.com/search?q=repo%3Agentoo%2Fgentoo+x11-libs%2Fgtk+gui-libs%2Fgtk&type=code

Comment (xen0n) [reply net-proxy/yass/yass-1.8.3.ebuild:29]: > Yes. It is the best solution I have ever found.
> 
> For loongarch64, it depends on gtk3 and for others, it depends on gtk4.

For the record you could do:

```
loong? ( x11-libs/gtk+:3=[wayland?] )
!loong? ( gui-libs/gtk:4=[wayland?] )
```

Take-away points:

* `ARCH` is also implicitly usable as USE flag
* Use sub-slot operator to ensure the package is correctly rebuilt as revdep

Comment (xen0n) [reply net-proxy/yass/yass-1.8.3.ebuild:29]: > > > Yes. It is the best solution I have ever found.
> > > For loongarch64, it depends on gtk3 and for others, it depends on gtk4.
> > 
> > For the record you could do:
> > 
> > ```
> > loong? ( x11-libs/gtk+:3=[wayland?] )
> > !loong? ( gui-libs/gtk:4=[wayland?] )
> > ```
> > 
> > Take-away points:
> > 
> > * `ARCH` is also implicitly usable as USE flag
> > * Use sub-slot operator to ensure the package is correctly rebuilt as revdep
> 
> Thanks for the comments. It also needs "and" operator on USE flags such as ``gui && loong? ( xxx )`` but I guess portage doesn't support that.

Just nest expressions in this case. You could find plenty of examples in various Gentoo repos such as the main tree or in this overlay for your reference.

Like:

```
gui? (
    loong? ( x11-libs/gtk+:3=[wayland?] )
    !loong? ( gui-libs/gtk:4=[wayland?] )
)
```

Comment (Chilledheart) [reply net-proxy/yass/yass-1.8.3.ebuild:29]: > > > > Yes. It is the best solution I have ever found.
> > > > For loongarch64, it depends on gtk3 and for others, it depends on gtk4.
> > > 
> > > 
> > > For the record you could do:
> > > ```
> > > loong? ( x11-libs/gtk+:3=[wayland?] )
> > > !loong? ( gui-libs/gtk:4=[wayland?] )
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Take-away points:
> > > 
> > > * `ARCH` is also implicitly usable as USE flag
> > > * Use sub-slot operator to ensure the package is correctly rebuilt as revdep
> > 
> > 
> > Thanks for the comments. It also needs "and" operator on USE flags such as `gui && loong? ( xxx )` but I guess portage doesn't support that.
> 
> Just nest expressions in this case. You could find plenty of examples in various Gentoo repos such as the main tree or in this overlay for your reference.

Cool. I will find some time to learn more about it.

**Pull request merged**